File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode start = new ListNode();
        ListNode final_start = start;
        int len = length(head);
        int pos = len - n + 1;
        ListNode temp = head;
        while (pos-- > 1) {
            start.next = temp;
            start = start.next;
            temp = temp.next;
        }
        start.next = temp.next;
        return final_start.next;
    }
    public int length(ListNode node) {
        int count = 0;
        while (node != null) {
            count++;
            node = node.next;
        }
        return count;
    }
}

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode()
        dummy.next = head
        def go_next(current_node):
            if current_node is None:
                return 0
            sub_ans = go_next(current_node.next)
            if sub_ans == n:
                current_node.next = current_node.next.next
            return 1 + sub_ans
        go_next(dummy)
        return dummy.next